### PR TITLE
Engine - Restructure KVDB Component for Enhanced Modularity and Testing

### PR DESCRIPTION
**Pull request by soynof**
|Related issue|
|---|
|#19341|

## Description

The KVDB component, currently a static library, is integral to various other components within Wazuh, including the Builder component for creating helper functions and the API component for exposing access to key-value databases (KVDB). To enhance modularity, maintainability, and testing, a restructuring of the KVDB component is proposed.

**Objectives:**

1. **Library Splitting:** Adapt KVDB to a new schema by splitting the library into interface and implementation. The C++ code for the interface is implemented, but it needs to be separated in cmake for better modularity.

2. **Test Relocation and Improvement:** Move the tests from the `/test` folder to `/kvdb/test`, categorizing them into unit tests and component tests. Improve the tests by parameterizing them for more comprehensive coverage.

3. **Mock Creation and Test Modification:** Develop Mocks for the interface and update the current API and builder tests to utilize these mocks. This will facilitate more robust and isolated testing of the KVDB component.

### CODE REVIEW

File: src/engine/source/api/CMakeLists.txt [hunk 1]
```diff
@@ -49,6 +49,8 @@ target_link_libraries(api
     fmt
     json
     kvdb
+    kvdb::ikvdb
+    kvdb::mocks
```
Comment (juliancnn) [inline src/engine/source/api/CMakeLists.txt:53]: Just need link kvdb::ikvdb

File: src/engine/source/api/test/src/unit/kvdb/kvdb_api_test.cpp [hunk 2]
```diff
@@ -1,25 +1,23 @@
...
 #include <fstream>
 #include <string>
 
+#include <gtest/gtest.h>
+
 #include <api/kvdb/handlers.hpp>
-#include <kvdb/kvdbManager.hpp>
+#include <kvdb/ikvdbhandler.hpp>
+#include <kvdb/mockKvdbHandler.hpp>
+#include <kvdb/mockKvdbManager.hpp>
```
Comment (juliancnn) [inline src/engine/source/api/test/src/unit/kvdb/kvdb_api_test.cpp:9]: Do not include:
```
#include <api/kvdb/handlers.hpp>
#include <kvdb/ikvdbhandler.hpp>
```

File: src/engine/source/kvdb/CMakeLists.txt [hunk 3]
```diff
@@ -22,19 +27,63 @@ CPMAddPackage(
...
+)
+
+target_include_directories(kvdb_ctest
+    PRIVATE
+    ${TEST_SRC_DIR})
+
+target_link_libraries(kvdb_ctest gtest_main metrics kvdb)
+gtest_discover_tests(kvdb_ctest)
+
+endif(ENGINE_BUILD_TEST)
```
Comment (juliancnn) [inline src/engine/source/kvdb/CMakeLists.txt:89]: add endline

File: src/engine/test/source/builder/CMakeLists.txt [hunk 4]
```diff
@@ -171,12 +171,18 @@ gtest_discover_tests(opBuilderHelperMap_test)
...
   ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/opBuilderKVDBSet_test.cpp
   ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/opBuilderKVDBGetArray_test.cpp
+  ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/opBuilderKVDBMatch_test.cpp
+  ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/opBuilderKVDBNotMatch_test.cpp
   # Use KVDB in builtime
   ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/opBuilderKVDBDecodeBitmask_test.cpp
   ${BUILDER_BUILDERS_TEST_SOURCE_DIR}/windowsHelper_test.cpp
 )
+
+target_link_libraries(opBuilderHelperKVDB_test kvdb_mocks test_mocks)
```
Comment (juliancnn) [inline src/engine/test/source/builder/CMakeLists.txt:184]:
```suggestion
target_link_libraries(opBuilderHelperKVDB_test kvdb::mocks)
```

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/source/api/test/src/unit/kvdb/kvdb_api_test.cpp (hunk 2 above)
Comment (soynof) [reply src/engine/source/api/test/src/unit/kvdb/kvdb_api_test.cpp:9]: The <api/kvdb/handlers.hpp> is used to test the api command, for example: managerGet, managerPost...

**Review by JavierBejMen — CHANGES_REQUESTED** [review]
Aside the comment, LGTM

File: src/engine/source/api/CMakeLists.txt (hunk 1 above)
Comment (juliancnn) [reply src/engine/source/api/CMakeLists.txt:53]: You should not link KVDB or kvdb::mocks, only the interface kvdb::ikvdb

**Review by juliancnn — CHANGES_REQUESTED** [review]

**Review by juliancnn — APPROVED** [review]
LGTM!

**Review by JavierBejMen — APPROVED** [review]

**Review by NahuFigueroa97 — APPROVED** [review]
LGTM!

**Pull request merged**